File: 141. Linked List Cycle/main.py
# Definition for singly-linked list.
from typing import Optional

class ListNode(object):
    def __init__(self, x: int) -> None:
        self.val: int = x
        self.next: Optional['ListNode'] = None

class Solution(object):
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        """
        :type head: ListNode
        :rtype: bool
        """
        # Floyd's cycle-finding algorithm
        if not head or not head.next:
            return False
        slow = head
        fast = head.next
        while slow != fast:
            if not fast or not fast.next:
                return False
            assert slow is not None
            assert fast is not None
            slow = slow.next
            fast = fast.next.next
        return True
            
        
    
# Test cases
if __name__ == "__main__":
    s = Solution()
    p = ListNode(3)
    p.next = ListNode(2)
    p.next.next = ListNode(0)
    p.next.next.next = ListNode(-4)
    p.next.next.next.next = p.next
    print(s.hasCycle(p))

File: 141. Linked List Cycle/main.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
function ListNode(val)
{
    this.val = val;
    this.next = null;

}
/**
 * @param {ListNode} head
 * @return {boolean}
 */

// Tortoise and Hare Algorithm
var hasCycle = function(head) {
    let slow = head
    let fast = head

    while (fast?.next != null) {
        slow = slow.next
        fast = fast.next.next

        if (slow == fast){
            return true
        }
    }
    return false

    
};

const list = new ListNode(1)
list.next = new ListNode(3)
list.next.next = new ListNode(4)
list.next.next.next = new ListNode(5)
list.next.next.next.next = list.next

console.log(hasCycle(list))